rust/otap-dataflow/crates/otap/src/otap_grpc/otlp/server_new.rs

```diff
@@ -0,0 +1,630 @@
+    pub(crate) Arc<Mutex<SlotsState<oneshot::Sender<Result<(), NackMsg<OtapPdata>>>>>>,
+);
+
+impl AckSlot {
+    /// Build a new per-signal slot map sized for the configured concurrency.
+    pub(crate) fn new(max_size: usize) -> Self {
+        Self(Arc::new(Mutex::new(SlotsState::new(max_size))))
+    }
+}
+
+/// The outcome from RouteResponse
+pub enum RouteResponse {
+    /// The Ack/Nack was sent.
+    Sent,
+    /// The Ack/Nack may have timed out.
+    Expired,
+    /// No subscription was found.
+    None,
+    /// The Ack/Nack had invalid call data.
+    Invalid,
+}
+
+impl AckSlot {
+    /// Routes the final outcome into the registered slot matching the provided `CallData`.
+    #[must_use]
+    pub fn route_response(
+        &self,
+        calldata: CallData,
+        result: Result<(), NackMsg<OtapPdata>>,
+    ) -> RouteResponse {
+        // Decode slot key from calldata
+        let key: SlotKey = match calldata.try_into() {
+            Ok(data) => data,
+            Err(_) => return RouteResponse::Invalid,
+        };
+
+        // Try to take the channel from the slot under the mutex.
+        let chan = self.0.lock().take(key);
+
+        // Try to send.
+        if chan.and_then(|sender| sender.send(result).ok()).is_some() {
+            RouteResponse::Sent
+        } else {
+            RouteResponse::Expired
+        }
+    }
+}
+
+/// Common settings for OTLP receivers.
+/// Per-signal server settings derived from user configuration and shared with the services.
+#[derive(Clone, Debug)]
+pub struct OtlpServerSettings {
+    /// Maximum concurrent requests per receiver instance (per core).
+    pub max_concurrent_requests: usize,
+    /// Whether the receiver should wait.
+    pub wait_for_result: bool,
+    /// Maximum size for inbound gRPC messages.
+    pub max_decoding_message_size: Option<usize>,
+    /// Request compression allowed
+    pub request_compression_encodings: EnabledCompressionEncodings,
+    /// Response compression used
+    pub response_compression_encodings: EnabledCompressionEncodings,
+}
+
+/// Precomputed empty responses per signal to avoid per-call prost encoding.
+fn precomputed_response(signal: SignalType) -> &'static [u8] {
+    static LOGS: OnceLock<Bytes> = OnceLock::new();
+    static METRICS: OnceLock<Bytes> = OnceLock::new();
+    static TRACES: OnceLock<Bytes> = OnceLock::new();
+
+    match signal {
+        SignalType::Logs => LOGS
+            .get_or_init(|| {
+                let mut buf = Vec::with_capacity(
+                    ExportLogsServiceResponse {
+                        partial_success: None,
+                    }
+                    .encoded_len(),
+                );
+                ExportLogsServiceResponse {
+                    partial_success: None,
+                }
+                .encode(&mut buf)
+                .expect("encode logs response");
+                Bytes::from(buf)
```

Comment:
I'll suggest applying this suggestion in a future PR: https://github.com/lquerel/otel-arrow/pull/5